csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -0,0 +1,149 @@
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="format">The Slice format (Compact or Sliced) used by the encapsulation.</param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame.</returns>
+        public static OutgoingRequestFrame Start(IObjectPrx proxy, string operation, bool idempotent,
+            FormatType? format = null, Context? context = null)
+        {
+            var frame = new OutgoingRequestFrame(proxy.Communicator, proxy.Identity, proxy.Facet, operation, idempotent,
+                proxy.Context, context);
+            frame.StartEncapsulation(proxy.EncodingVersion, format);
+            return frame;
+        }
+
+        /// <summary>Creates a new outgoing request frame with for in-parameters payload the in-parameters of an
+        /// incoming request frame.</summary>
+        /// <param name="proxy">A proxy to the target Ice object. This method uses the communicator, identity, facet
+        /// and context of this proxy to create the request frame.</param>
+        /// <param name="operation">The operation to invoke on the target Ice object.</param>
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="payload">The payload of this request frame, which represents the marshaled in-parameters.
+        /// </param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame.</returns>
+        public static OutgoingRequestFrame Create(IObjectPrx proxy, string operation, bool idempotent,
+            ArraySegment<byte> payload, Context? context = null)
+        {
+            var frame = new OutgoingRequestFrame(proxy.Communicator, proxy.Identity, proxy.Facet, operation, idempotent,
+                proxy.Context, context);
+
+            if (payload.Count == 0)
+            {
+                frame.WriteEmptyEncapsulation(proxy.EncodingVersion);
+            }
+            else
+            {
+                // TODO: works only because we know how payload is created!
+                frame.WriteBlob(payload.Array);
+            }
+            frame.IsSealed = true;
+            return frame;
+        }
+
+         /// <summary>Marks the end of the in-parameters for this request frame.</summary>
+        public void EndPayload()
+        {
+            Debug.Assert(!IsSealed); // TODO: throw an exception
+            EndEncapsulation();
+            IsSealed = true;
+            // Note: with the 2.0 encoding the frame will be sealed later, after the context is written.
+        }
```

Comment:
We currently use the API like:

```csharp
var frame = global::Ice.OutgoingResponseFrame.Start(...)
frame.write(...);
frame.endPayload();
```

Could this instead be  replaced by a passing a delete/lambda to do the writing the `OutgoingRequestFrame` constructor.

```csharp
Ice.OutgoingResponseFrame(otherParams..., Action<Ice.OutputStream> callback)
{
    // other setup
    ...
    start();
    callback(this);
    end();  
}
```

Now the user of `Ice.OutgoingResponseFrame` doesn't have to deal with `start()/end()`

Reply:
Fixed as suggested. It makes the API nicer at the cost of a delegate call for each request/response marshal.

Reply:
With the new OutgoingXxxFrame ctors, the parameter order is consistent:
IObjectPrx, ..., Context?, [Format?, ]payload
or
Current, [Format?, ]payload

The format is only meaningful when the payload is to be marshaled.